clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -376,7 +390,7 @@ public boolean isClosed() {
     }
 
     public boolean isFull() {
-        return isClosed() || this.writeLimit <= estimatedBytesWritten();
+        return isClosed() || (this.numRecords > 0 && this.writeLimit <= estimatedBytesWritten());
```

Comment:
Yes, correct. The difference from the old code is really in the `hasRoomFor` method, which now uses `isFull()`.